facet-macros-emit/src/derive.rs

```diff
@@ -116,3 +116,54 @@ pub(crate) fn build_type_params(generics: Option<&GenericParams>) -> TokenStream
         quote! { .type_params(&[#(#type_params),*]) }
     }
 }
+
+/// Generate the `type_name` function for the `ValueVTable`,
+/// displaying realized generics if present.
+pub(crate) fn generate_type_name_fn(
+    type_name: &Ident,
+    generics: Option<&GenericParams>,
+) -> TokenStream {
+    let type_name_str = type_name.to_string();
+
+    let write_generics = generics.and_then(|generics| {
+        let params = generics.params.0.iter();
+        let write_each = params.filter_map(|param| match &param.value {
+            // Lifetimes not shown by `std::any::type_name`, this is parity.
+            GenericParam::Lifetime { .. } => None,
+            GenericParam::Const { name, .. } => Some(quote! {
+                write!(f, "{:?}", #name)?;
+            }),
+            GenericParam::Type { name, .. } => Some(quote! {
+                <#name as ::facet::Facet>::SHAPE.vtable.type_name()(f, opts)?;
+            }),
+        });
+        // TODO: is there a way to construct a DelimitedVec from an iterator?
```

Comment:
Have you seen this?

https://github.com/cehteh/unsynn/blob/236a9f29d0f52c680093645317134c6cc00f6aca/src/container.rs#L365-L393

Reply:
https://github.com/cehteh/unsynn/commit/3d4aa5af009cf7af5722192cd35645244ee4c49d hasn't appeared in a release yet, let's leave the TODO there.